vignettes/namespaces.Rmd

```diff
@@ -0,0 +1,288 @@
+prefix and carry around that object. Note: an `xml_namespace` object is a named
+character vector, so we can create it with `structure()` and use it to introduce
+semantically sensible XPath queries
+
+```{r}
+ns <- structure(c(md = "http://commonmark.org/xml/1.0"), class = "xml_namespace")
+xml2::xml_find_all(xml, "//md:text", ns)
+```
+
+Of course, now if we want to make any semantic XPath query, we need to include
+both a prefix and a namespace object.
+
+### Transformers!
+
+The commonmark namespace allows us to transform our document to markdown using 
+an XSLT stylesheet, which is---that's right---an XML document:
+
+```{r stysh}
+sty <- xml2::read_xml(tinkr::stylesheet())
+sty
+```
+
+Each `xsl:template` node in this stylesheet matches against a specific node in
+the commonmark namespace (prefix: `md`) and emits text based on that node. This
+allows us to write back to markdown:
+
+```{r}
+cat(xslt::xml_xslt(xml, sty))
+```
+
+We can in this way programatically transform the content of the markdown:
+
+```{r}
+xml <- commonmark::markdown_xml("this is a **test**") %>%
+  xml2::read_xml()
+
+xml2::xml_find_all(xml, "//md:strong", ns) %>%
+  xml2::xml_set_name("code") %>%
+  xml2::xml_set_text("r cat('_test_')")
+
+sty <- xml2::read_xml(tinkr::stylesheet())
+cat(xslt::xml_xslt(xml, sty))
+```
+
+### Perils: adding nodes
+
+A default namespace is all fun and games until you need to add new nodes. Take
+for example the situation where we want to add a code block. In commonmark, it's
+a `code_block` node with an `info` attribute stating the language and the text
+inside is the code.
+
+```{r}
+xml <- commonmark::markdown_xml("this is a **test**") %>%
+  xml2::read_xml() 
+xml2::xml_add_child(xml, "code_block", info = "{r}", "1 + rnorm(1)\n")
+xml
+xml2::xml_find_all(xml, "//md:code_block", ns)
+sty <- xml2::read_xml(tinkr::stylesheet())
+cat(xslt::xml_xslt(xml, sty))
+```
+
+By all means, the node should have added correctly, but because we did not 
+specify a namespace, *it is not recognized as part of the `md` namespace* even
+though we added it as a child of the document. The [best way to handle this
+situation is to reparse the document][reparse]:
+
+```{r reread}
+xml %>%
+  as.character() %>%
+  xml2::read_xml() %>%
+  xslt::xml_xslt(sty) %>%
+  cat()
+```
+
+We *could* also try adding the namespace to the node when we add it:
+
+```{r}
+xml <- commonmark::markdown_xml("this is a **test**") %>%
+  xml2::read_xml() 
+xml2::xml_add_child(xml, "code_block", 
+  xmlns = "http://commonmark.org/xml/1.0", info = "{r}", "1 + rnorm(1)\n")
+xml
+xml2::xml_find_all(xml, "//md:code_block", ns)
+cat(xslt::xml_xslt(xml, sty))
+```
+
+It works, but let's take a look at our namespaces:
+
+```{r}
+xml2::xml_ns(xml)
+```
+
+Every node we add with an unnamed namespace adds another default and in the end, 
+if we are doing a lot of substitution, we can end up with hundreds of namespaces.
```

Comment:
:scream:

Reply:
Yeah, this was one of the horrors I stumbled upon when writing {pegboard}.